compiler/src/iree/compiler/Codegen/Common/LinkTuningSpecsPass.cpp

```diff
@@ -136,6 +237,128 @@ emitLinkedTuningSpec(ModuleOp module, ArrayRef<NamedSequenceOp> specsToLink) {
   return newSpec;
 }
 
+static FailureOr<NamedSequenceOp> emitLinkedDefaultTuningSpec(ModuleOp module) {
+  OpBuilder builder(module.getContext());
+  SmallVector<NamedSequenceOp> namedSequenceOpsToMove;
+  SmallVector<ForeachMatchOp> foreachMatchOps;
+  llvm::DenseMap<NamedSequenceOp, ForeachMatchOp> namedSequenceToForeachMatch;
+
+  // Step 1: Collect NamedSequenceOps and ForeachMatchOps from inner modules.
+  for (auto innerModule : module.getBody()->getOps<ModuleOp>()) {
+    collectNamedSequenceAndForeachMatchOpsFromModule(
+        innerModule, namedSequenceOpsToMove, foreachMatchOps,
+        namedSequenceToForeachMatch);
+  }
+
+  // Step 2-a: Make sure the name sequence names are unique, and then move
+  // collected NamedSequenceOps to the top-level module.
+  Type anyOpType = builder.getType<transform::AnyOpType>();
+  llvm::StringMap<unsigned> specNameCounts;
+  for (NamedSequenceOp op : namedSequenceOpsToMove) {
+    StringRef specName = op.getSymName();
+    std::string newSpecName = getUniqueSpecName(specName, specNameCounts);
+    op.setSymName(newSpecName);
+
+    // Skip updating ForeachMatchOp if the NamedSequenceOp is not used in it
+    // or its name has not changed.
+    if (!namedSequenceToForeachMatch.contains(op) || newSpecName == specName) {
+      op.getOperation()->moveBefore(module.getBody(), module.getBody()->end());
+      continue;
+    }
+
+    ForeachMatchOp foreachMatchOp = namedSequenceToForeachMatch[op];
+
+    SmallVector<Attribute> updatedMatchers, updatedActions;
+    for (auto matcherAttr : foreachMatchOp.getMatchers()) {
+      StringRef matcherName =
+          cast<SymbolRefAttr>(matcherAttr).getRootReference();
+      updatedMatchers.push_back(
+          (matcherName == specName)
+              ? SymbolRefAttr::get(builder.getContext(), newSpecName)
+              : matcherAttr);
+    }
+
+    for (auto actionAttr : foreachMatchOp.getActions()) {
+      StringRef actionName = cast<SymbolRefAttr>(actionAttr).getRootReference();
+      updatedActions.push_back(
+          (actionName == specName)
+              ? SymbolRefAttr::get(builder.getContext(), newSpecName)
+              : actionAttr);
```

Comment:
I'd take this logic out of the push back to make debugging easier. It's simpler to print a variable than to print an element in a vector